sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+                                                    new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      return updateCurrent(searchResult);
+    }
+
+    private boolean updateCurrent(JsonObject searchResult) {
+      //stop if no more data
+      if (searchResult.getAsJsonObject("hits").getAsJsonArray("hits").size() == 0) {
+        return false;
+      }
+      current = searchResult.getAsJsonObject("hits").getAsJsonArray("hits").get(0)
+          .getAsJsonObject().getAsJsonObject("_source").toString();
+      return true;
+    }
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      return current;
+    }
+
+    @Override
+    public void close() throws IOException {
+      // remove the scroll
+      String requestBody = String.format("{\"scroll_id\" : [\"%s\"]}", scroll);
+      HttpEntity entity = new NStringEntity(requestBody,
+                                            ContentType.APPLICATION_JSON);
+      try {
+        restClient.performRequest("DELETE", "/_search/scroll",
+                                  Collections.<String, String>emptyMap(),
+                                  entity, new BasicHeader("", ""));
+      } catch (IOException e) {
+        //ES gives the same scroll id to each source, close that scroll more than once produces an
+        // error is ES 5.x so protect this: cannot request if a particular scroll id is still open
+        // so ignores exception (raised with ES v5.x)
+        if (e.getMessage() != null && !e.getMessage().contains("no scroll ids specified")) {
+          throw e;
+        }
+      }
+      // close client
+      if (restClient != null) {
+        restClient.close();
+      }
+    }
+
+    @Override
+    public BoundedSource<String> getCurrentSource() {
+      return source;
+    }
+  }
+
+  /**
+   * A {@link PTransform} writing data to Elasticsearch.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<String>, PDone> {
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    abstract long getBatchSize();
+    abstract int getBatchSizeMegaBytes();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+      abstract Builder setBatchSize(long batchSize);
+      abstract Builder setBatchSizeMegaBytes(int batchSizeMegaBytes);
+      abstract Write build();
+    }
+
+    public Write withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.write()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
```

Comment:
I added doc in the class javadoc not on the method. But I will add javadoc for every with* end user method